app/Main.hs

```diff
@@ -162,24 +162,83 @@ runMain options = do
 
   -- Start a main event loop for every project.
   let
-    -- TODO: This is very, very ugly. Get these per-project collections sorted
-    -- out.
     zipped = zip4 (Config.projects config) projectQueues stateVars projectStates
-    tuples = map (\(cfg, (_, a), (_, b), (_, c)) -> (cfg, a, b, c)) zipped
-  projectThreads <- forM tuples $ \ (projectConfig, projectQueue, stateVar, projectState) -> do
+    tuples = map (\(cfg, (_, a), (_, b), (_, c)) -> ProjectThreadData cfg a b c) zipped
+  metrics <- Metrics.registerProjectMetrics
+  projectThreads <- forM tuples $ projectThread config options metrics
+
+  let
+    -- When the webhook server receives an event, enqueue it on the webhook
+    -- event queue if it is not full.
+    ghTryEnqueue = Github.tryEnqueueEvent ghQueue
+
+    -- Allow the webinterface to retrieve the latest project state per project.
+    getProjectState projectInfo =
+      fmap Logic.readStateVar $ lookup projectInfo stateVars
+    getOwnerState :: Owner -> IO [(ProjectInfo, ProjectState)]
+    getOwnerState owner = do
+      let states = filter (\(projectInfo, _) -> Project.owner projectInfo == owner) stateVars
+      mapM (\(info, state) ->  Logic.readStateVar state >>= \sVar -> pure (info, sVar)) states
+
+  let
+    port      = Config.port config
+    tlsConfig = Config.tls config
+    secret    = Config.secret config
+    -- TODO: Do this in a cleaner way.
+    infos     = getProjectInfo <$> Config.projects config
+  putStrLn $ "Listening for webhooks on port " ++ show port ++ "."
+  runServer <- fst <$> buildServer port tlsConfig infos secret ghTryEnqueue getProjectState getOwnerState
+  serverThread <- Async.async runServer
+  metricsThread <- runMetricsThread config
+
+  -- Note that a stop signal is never enqueued. The application just runs until
+  -- until it is killed, or until any of the threads stop due to an exception.
+  void $ Async.waitAny $ [serverThread, ghThread] ++ metricsThread ++ projectThreads
+
+data ProjectThreadData = ProjectThreadData
+  { projectThreadConfig   :: Config.ProjectConfiguration
+  , projectThreadQueue    :: Logic.EventQueue
+  , projectThreadStateVar :: Logic.StateVar
+  , projectThreadState    :: ProjectState
+  }
+
+projectThread :: Configuration
```

Comment:
Thanks for splitting up this function from the main function. That makes the code more readable!